tpke/src/refresh.rs

```diff
@@ -0,0 +1,164 @@
+    // Each participant prepares an update for each other participant
+    participants
+        .iter()
+        .map(|p1| {
+            let i = p1.index;
+            // Generate a new random polynomial with constant term 0
+            let d_i = make_random_polynomial::<E>(threshold, x_r, rng);
+
+            // Now, we need to evaluate the polynomial at each of participants' indices
+            let deltas_i: HashMap<_, _> =
+                compute_polynomial_deltas::<E>(participants, &d_i);
+            (i, deltas_i)
+        })
+        .collect::<HashMap<_, _>>()
+}
+
+fn update_shares_for_recovery<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    deltas: &HashMap<usize, HashMap<usize, E::G2Projective>>,
+) -> Vec<E::G2Projective> {
+    // TODO: Refactor this function so that each participant performs it individually
+    participants
+        .iter()
+        .map(|p| {
+            let i = p.index;
+            let mut new_y = E::G2Projective::from(
+                p.private_key_share.private_key_shares[0], // y_i
+            );
+            for j in deltas.keys() {
+                new_y += deltas[j][&i];
+            }
+            new_y
+        })
+        .collect()
+}
+
+fn make_random_polynomial<E: PairingEngine>(
+    threshold: usize,
+    x_r: &E::Fr,
+    rng: &mut impl RngCore,
+) -> Vec<E::Fr> {
+    // [][threshold-1]
+    let mut d_i = (0..threshold - 1)
+        .map(|_| E::Fr::rand(rng))
+        .collect::<Vec<_>>();
+    // [0..][threshold]
+    d_i.insert(0, E::Fr::zero());
+
+    // Now, we calculate d_i_0
+    // This is the term that will "zero out" the polynomial at x_r, d_i(x_r) = 0
+    let d_i_0 = E::Fr::zero() - evaluate_polynomial::<E>(&d_i, x_r);
+    d_i[0] = d_i_0;
+    assert_eq!(evaluate_polynomial::<E>(&d_i, x_r), E::Fr::zero());
+
+    assert_eq!(d_i.len(), threshold);
+
+    d_i
+}
+
+fn evaluate_polynomial<E: PairingEngine>(
+    polynomial: &[E::Fr],
+    x: &E::Fr,
+) -> E::Fr {
+    let mut result = E::Fr::zero();
+    let mut x_power = E::Fr::one();
+    for coeff in polynomial {
+        result += *coeff * x_power;
+        x_power *= x;
+    }
+    result
+}
+
+fn prepare_share_updates_for_refreshing<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    threshold: usize,
+    rng: &mut impl RngCore,
+) -> HashMap<usize, E::G2Projective> {
+    let coeffs = make_random_polynomial::<E>(threshold, &E::Fr::zero(), rng);
+    compute_polynomial_deltas(participants, &coeffs)
+}
+
+fn compute_polynomial_deltas<E: PairingEngine>(
+    participants: &[PrivateDecryptionContextSimple<E>],
+    coeffs: &[E::Fr],
+) -> HashMap<usize, E::G2Projective> {
+    participants
+        .iter()
+        .map(|p| {
+            let i = p.index;
+            let x_i = p.public_decryption_contexts[i].domain;
+            let eval = evaluate_polynomial::<E>(coeffs, &x_i);
+            let h_g2 = E::G2Projective::from(p.h);
```

Comment:
This could be taken outside of the loop, or even moved to a public context, since it's just the H generator in projective representation.

Reply:
In `tpke`, public context corresponds to variables shared with other participants, and since they already have `h` there is no reason to put it there. 

The different structs in `tpke` are going to be refactored so that should bring a bit more clarity.